### PR TITLE
ROX-19561: Add Sensor message rate limit

### PR DESCRIPTION
**Pull request by mtodor**
## Description

This PR adds global rate limiting for sensor messages sent from all connected secured clusters.

- Added env variable to set rate limit for sensor messages (check CHANGELOG)
- Refactored init sync manager to rate limit manager
  - Add support for Sensor message rate limiting
  - Add init sync boost (not configurable) per sensor
  - Changed previous logic for max init sync sensors - if it's unlimited, it will be set to `math.MaxInt` (that simplifies logic with boost)
- By default, we will not limit sensor messages, and for a negative rate limit, the app will panic on start

For benchmarks, check the description on PR: #8116

**Considerations**

1. Should we increase `boostInitSyncRateLimit`?
2. Should we remove boost logic?

**Pre-requirement:** #8116 

## Checklist
- [x] Investigated and inspected CI test results
- [x] Unit test and regression tests added
- [x] Evaluated and added CHANGELOG entry if required
- ~[ ] Determined and documented upgrade steps~
- [ ] Documented user facing changes (create PR based on [openshift/openshift-docs](https://github.com/openshift/openshift-docs) and merge into [rhacs-docs](https://github.com/openshift/openshift-docs/tree/rhacs-docs))

### Here I tell how I validated my change

1. Added GoLang unit tests
2. **Planning** to test it with `scale` scripts

### Reminder for reviewers

In addition to reviewing code here, reviewers **must** also review testing and request further testing in case the
performed one does not seem sufficient. As a reviewer, you must not approve the change until you understand the
performed testing and you are satisfied with it.

### CODE REVIEW

File: central/sensor/service/connection/ratelimit_manager.go [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+func (m *rateLimitManager) Add(clusterID string) bool {
+	m.mutex.Lock()
+	defer m.mutex.Unlock()
+
+	if len(m.initSyncSensors) >= m.maxSensors {
+		return false
+	}
+
+	if m.initSyncSensors.Add(clusterID) {
+		m.eventRateLimiter.IncreaseLimit(boostInitSyncRateLimit)
```
[the remark this answers is not in the document]
Comment (janisz) [reply central/sensor/service/connection/ratelimit_manager.go:59]: We can keep it as is to have simple code. IncreaseLimit should be fast enough so there is no real speed difference.

File: central/sensor/service/connection/ratelimit_manager_test.go [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+	t.Setenv(env.CentralSensorMaxEventsPerSecond.EnvVar(), "0")
+	m := NewRateLimitManager()
+
+	for i := 0; i < 100; i++ {
+		require.False(t, m.LimitMsg(), "No limit")
+	}
+}
+
+func TestNewRateLimitManagerEventsPerSecond(t *testing.T) {
+	t.Setenv(env.CentralSensorMaxEventsPerSecond.EnvVar(), "10")
```
Comment (janisz) [inline central/sensor/service/connection/ratelimit_manager_test.go:86]:
```suggestion
	t.Setenv(env.CentralSensorMaxEventsPerSecond.EnvVar(), "3")
```

**Review by janisz — APPROVED** [review]

**Review by rhybrillou — APPROVED** [review]

**Comment by connorgorman** [comment]
If there is a single cluster misbehaving, does it make sense to target it and throttle that? I'm not sure that a single high cap for all sensors gives us the flexibility we want

**Comment by mtodor** [comment]
@connorgorman I understand your concerns, and I agree that the flexibility here is limited. In the case of ACSCS, our intention is to utilize this feature sparingly and only in exceptional circumstances as a manual intervention (SOP). Currently, our approach doesn't involve throttling messages; instead, it leads to the termination of a connection.

The primary objective of this limitation is to safeguard the central from resource depletion, whether caused by a single misbehaving sensor or a substantial load from multiple sensors. Therefore, the rate limit must be defined globally for the central.

With this requirement in mind, what solution would you suggest? One possibility is to track the "resource consumption" (number of messages) by each sensor and disconnect the largest consumer when the limit is reached.

**Comment by mtodor** [comment]
We decided to postpone this functionality and implement a solution that would isolate problematic sensors in a better way. I have created another ticket for that. I'll close this PR to keep the JIRA state clean.

**Pull request closed without merging**